iothub/device/samples/convention-based-samples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -0,0 +1,375 @@
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class TemperatureControllerSample
+    {
+        private const string Thermostat1 = "thermostat1";
+        private const string Thermostat2 = "thermostat2";
+
+        private static readonly Random s_random = new Random();
+
+        private readonly DeviceClient _deviceClient;
+        private readonly ILogger _logger;
+
+        // Dictionary to hold the temperature updates sent over each "Thermostat" component.
+        // NOTE: Memory constrained devices should leverage storage capabilities of an external service to store this
+        // information and perform computation.
+        // See https://docs.microsoft.com/en-us/azure/event-grid/compare-messaging-services for more details.
+        private readonly Dictionary<string, Dictionary<DateTimeOffset, double>> _temperatureReadingsDateTimeOffset =
+            new Dictionary<string, Dictionary<DateTimeOffset, double>>();
+
+        // Dictionary to hold the current temperature for each "Thermostat" component.
+        private readonly Dictionary<string, double> _temperature = new Dictionary<string, double>();
+
+        // Dictionary to hold the max temperature since last reboot, for each "Thermostat" component.
+        private readonly Dictionary<string, double> _maxTemp = new Dictionary<string, double>();
+
+        public TemperatureControllerSample(DeviceClient deviceClient, ILogger logger)
+        {
+            _deviceClient = deviceClient ?? throw new ArgumentNullException($"{nameof(deviceClient)} cannot be null.");
+            _logger = logger ?? LoggerFactory.Create(builer => builer.AddConsole()).CreateLogger<TemperatureControllerSample>();
+        }
+
+        public async Task PerformOperationsAsync(CancellationToken cancellationToken)
+        {
+            // This sample follows the following workflow:
+            // -> Set handler to receive and respond to writable property update requests.
+            // -> Set handler to receive and respond to commands.
+            // -> Update device information on "deviceInformation" component.
+            // -> Send initial device info - "workingSet" over telemetry, "serialNumber" over reported property update - root interface.
+            // -> Periodically send "temperature" over telemetry - on "Thermostat" components.
+            // -> Send "maxTempSinceLastReboot" over property update, when a new max temperature is set - on "Thermostat" components.
+
+            _logger.LogDebug("Subscribe to writable property updates.");
+            await _deviceClient.SubscribeToWritablePropertiesEventAsync(HandlePropertyUpdatesAsync, null, cancellationToken);
+
+            _logger.LogDebug($"Subscribe to commands.");
+            await _deviceClient.SubscribeToCommandsAsync(HandleCommandsAsync, null, cancellationToken);
+
+            await UpdateDeviceInformationAsync(cancellationToken);
```

Comment:
Seems on start up, the client should get twin to see what the latest writable setting values are. Or does that happen automatically with a callback?

Reply:
No, the sample does not have that logic implemented right now. If you think it is worth adding here, we could add it into the serial number update flow, just to demonstrate how to get the current property value and update only if required.
This will be a non-writable (from service perspective) property update though. Writable property updated are currently only executed through the callback (and are always executed). We could add a similar check there as well, if we want.

Reply:
I've added the check for the serial number flow, the target temperature (writable) flow will be similar, so I didn't add it there again.

Reply:
When we don't follow best practices in our samples, and then wonder why our customers do things we don't suggest, we shouldn't be surprised. Let's stick to best practices.